android/java/org/chromium/chrome/browser/playlist/PlaylistHostActivity.java

```diff
@@ -0,0 +1,492 @@
+            if (playlistId.equals(ConstantUtils.ALL_PLAYLIST)) {
+                showAllPlaylistsFragment();
+            } else {
+                showPlaylist(playlistId, false);
+            }
+        }
+    }
+
+    private void loadPlaylist(String playlistId) {
+        if (mPlaylistService != null) {
+            mPlaylistService.getPlaylist(playlistId, playlist -> {
+                JSONObject playlistJson = new JSONObject();
+                try {
+                    playlistJson.put("id", playlist.id);
+                    playlistJson.put("name", playlist.name);
+                    JSONArray playlistItemsJsonArray = new JSONArray();
+                    for (PlaylistItem playlistItem : playlist.items) {
+                        JSONObject playlistItemObject = new JSONObject();
+                        playlistItemObject.put("id", playlistItem.id);
+                        playlistItemObject.put("name", playlistItem.name);
+                        playlistItemObject.put("page_source", playlistItem.pageSource.url);
+                        playlistItemObject.put("media_path", playlistItem.mediaPath.url);
+                        playlistItemObject.put("media_src", playlistItem.mediaSource.url);
+                        playlistItemObject.put("thumbnail_path", playlistItem.thumbnailPath.url);
+                        playlistItemObject.put("cached", playlistItem.cached);
+                        playlistItemObject.put("author", playlistItem.author);
+                        playlistItemObject.put("duration", playlistItem.duration);
+                        playlistItemObject.put(
+                                "last_played_position", playlistItem.lastPlayedPosition);
+                        playlistItemsJsonArray.put(playlistItemObject);
+                    }
+                    playlistJson.put("items", playlistItemsJsonArray);
+                    if (mPlaylistViewModel != null) {
+                        mPlaylistViewModel.setPlaylistData(playlistJson.toString(2));
+                    }
+                } catch (Exception e) {
+                    Log.e(TAG, "PlaylistHostActivity -> JSONException error " + e);
+                }
+            });
+        }
+    }
+
+    private void showPlaylist(String playlistId, boolean shouldFallback) {
+        loadPlaylist(playlistId);
+        PlaylistFragment playlistFragment = new PlaylistFragment();
+        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction().replace(
+                R.id.fragment_container_view_tag, playlistFragment);
+        if (shouldFallback) {
+            transaction.addToBackStack(AllPlaylistFragment.class.getSimpleName());
+        }
+        transaction.commit();
+    }
+
+    private void loadAllPlaylists() {
+        if (mPlaylistService != null) {
+            mPlaylistService.getAllPlaylists(playlists -> {
+                try {
+                    JSONArray playlistsJson = new JSONArray();
```

Comment:
does it make sense to form JSON on the main thread? JSON parsers are usually slow especially on big files. Could it be pretty big? If a user has 500 items for example?